Allow override of backend in dist.new_group()

pietern opened the pull request:
Summary:
There is no need to force the backend to be the same as the global
process group, as long as the backend is "nccl" or "gloo".

Differential Revision: D14657204

apaszke commented on torch/distributed/distributed_c10d.py:
```diff
@@ -230,20 +230,28 @@ def _check_tensor_list(param, param_name):
+    Checks if the Gloo backend is available.
+
+    """
+    return True
```
This seems wrong? IIRC you can build PyTorch without gloo, so we should have a check for that

pietern replied:
@apaszke You can, but right now it won't build c10d, so it won't load `torch.distributed` at all. This should be fix in and by itself and then this function can be updated as well (added #18851).

pietern commented:
@apaszke Updated `is_gloo_available` to do the right thing regardless of build configuration.

Added docstring for new `backend` kwarg to `new_group`.

pietern committed:
Allow override of backend in dist.new_group() (#18595)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18595

There is no need to force the backend to be the same as the global
process group, as long as the backend is "nccl" or "gloo".

Reviewed By: mrshenli

Differential Revision: D14657204

fbshipit-source-id: d8764874e8cfc5306c00e1ca9a951ae88edb7875

densechen commented:
Is there any demonstration of the usage of new_group()?
